src/main/java/com/rtg/vcf/eval/VcfEvalCli.java

```diff
@@ -400,29 +403,84 @@ protected VcfEvalParams makeParams() throws IOException {
-    if (flags.isSet(VcfEvalCli.ROC_SUBSET)) {
-      final List<?> values = flags.getValues(VcfEvalCli.ROC_SUBSET);
-      for (Object o : values) {
-        rocFilters.add(((RocFilterProxy) o).filter());
-      }
-    }
+    final LinkedHashMap<String, Set<RocFilter>> rocFiltersTypeMap = new LinkedHashMap<>();
+
+    final Set<RocFilter> rocExprFilters = new LinkedHashSet<>();
     if (flags.isSet(ROC_EXPR)) {
       final List<?> values = flags.getValues(VcfEvalCli.ROC_EXPR);
       for (Object o : values) {
         final String[] e = StringUtils.split((String) o, '=', 2);
         assert e.length == 2;
-        rocFilters.add(new ExpressionRocFilter(e[0], e[1]));
+        rocExprFilters.add(new ExpressionRocFilter(e[0], e[1]));
       }
     }
+    rocFiltersTypeMap.put(ROC_EXPR, rocExprFilters);
+
+    final Set<RocFilter> rocRegionsFilters = new LinkedHashSet<>();
     if (flags.isSet(ROC_REGIONS)) {
       final List<?> values = flags.getValues(VcfEvalCli.ROC_REGIONS);
       for (Object o : values) {
         final String[] e = StringUtils.split((String) o, '=', 2);
         assert e.length == 2;
-        rocFilters.add(new RegionsRocFilter(e[0], new File(e[1])));
+        rocRegionsFilters.add(new RegionsRocFilter(e[0], new File(e[1])));
+      }
+    }
+    rocFiltersTypeMap.put(ROC_REGIONS, rocRegionsFilters);
+
+    // Split subset filters into genotype or variant flavors for intersecting in ROC_CROSS_JOIN
+    final Set<RocFilter> rocGenotypeFilters = new LinkedHashSet<>();
+    final Set<RocFilter> rocVariantFilters = new LinkedHashSet<>();
+    if (flags.isSet(VcfEvalCli.ROC_SUBSET)) {
+      final List<?> values = flags.getValues(VcfEvalCli.ROC_SUBSET);
+      final Set<String> genotypeOptions = new HashSet<>(Arrays.asList(RocFilter.HET.name(), RocFilter.HOM.name()));
+      final Set<String> variantOptions = new HashSet<>(
+              Arrays.asList(RocFilter.SNP.name(), RocFilter.NON_SNP.name(), RocFilter.INDEL.name(), RocFilter.MNP.name())
+      );
+      for (Object o : values) {
+        RocFilter rocFilter = ((RocFilterProxy) o).filter();
+        if (genotypeOptions.contains(rocFilter.name())) {
+          rocGenotypeFilters.add(rocFilter);
+        } else if (variantOptions.contains(rocFilter.name())) {
+          rocVariantFilters.add(rocFilter);
+        } else {
+          // exception
+        }
       }
     }
+    rocFiltersTypeMap.put("rocGenotypeFilters", rocGenotypeFilters);
+    rocFiltersTypeMap.put("rocVariantFilters", rocVariantFilters);
+
+    final Set<RocFilter> rocFilters = new LinkedHashSet<>();
+    if (flags.isSet(ROC_CROSS_JOIN)) {
+      // Add decoy "ALL" filters to ensure following cartesian product isn't empty and hit every combination
+      rocFiltersTypeMap.get(ROC_EXPR).add(RocFilter.ALL);
+      rocFiltersTypeMap.get(ROC_REGIONS).add(RocFilter.ALL);
+      rocFiltersTypeMap.get("rocGenotypeFilters").add(RocFilter.ALL);
+      rocFiltersTypeMap.get("rocVariantFilters").add(RocFilter.ALL);
```

Comment:
Perhaps we should only add these `RocFilter.ALL` entries to the maps if they are otherwise empty.

If the user gives `--roc-cross-join --roc-subset het,hom,snp` I think I would expect the following output files:
- weighted_roc.tsv (which we always produce)
- het+snp_roc.tsv
- hom+snp_roc.tsv

I don't think I would expect to see:
- snp_roc.tsv
- het_roc.tsv
- hom_roc.tsv

If the user gives `--roc-cross-join --roc-subset het,hom,snp --roc-expr "test=INFO.TEST>5" --roc-regions "mybed=mybed.bed"` I think I would expect the following output files:
- weighted_roc.tsv (which we always produce)
- test+mybed+het+snp_roc.tsv
- test+mybed+hom+snp_roc.tsv

I don't think I would expect to see:
- snp_roc.tsv
- het_roc.tsv
- hom_roc.tsv
- test_roc.tsv
- mybed_roc.tsv
- test+mybed_roc.tsv
- test+snp_roc.tsv

Agree?
- test+het_roc.tsv
- test+hom_roc.tsv
- mybed+het+snp_roc.tsv
- mybed+hom+snp_roc.tsv
- mybed+het_roc.tsv
- mybed+hom_roc.tsv